tests/service_atcoder.py

```diff
@@ -303,6 +304,187 @@ def test_from_html(self):
+        resp = requests.get(url)
+        html = resp.content.decode(resp.apparent_encoding)
+        data = AtCoderProblemDetailedData.from_html(html, problem=AtCoderProblem.from_url(url))
+
+        self.assertEqual(data.alphabet, 'D')
+        self.assertEqual(data.available_languages, None)
+        self.assertEqual(data.html, html)
+        self.assertEqual(data.input_format, '\r\n<var>R</var> <var>C</var>\r\n<var>X</var> <var>Y</var>\r\n<var>D</var> <var>L</var>\r\n')
+        self.assertEqual(data.memory_limit_byte, 64 * 1000 * 1000)
+        self.assertEqual(data.name, 'AtCoder社の冬')
+        self.assertEqual(data.problem, AtCoderProblem.from_url(url))
+        self.assertEqual(data.sample_cases, [
+            TestCase(name='sample-1', input_name='入力例 1', input_data=b'3 2\n2 2\n2 2\n', output_name='出力例 1', output_data=b'12\n'),
+            TestCase(name='sample-2', input_name='入力例 2', input_data=b'4 5\n3 1\n3 0\n', output_name='出力例 2', output_data=b'10\n'),
+            TestCase(name='sample-3', input_name='入力例 3', input_data=b'23 18\n15 13\n100 95\n', output_name='出力例 3', output_data=b'364527243\n'),
+            TestCase(name='sample-4', input_name='入力例 4', input_data=b'30 30\n24 22\n145 132\n', output_name='出力例 4', output_data=b'976668549\n'),
+        ])
+        self.assertEqual(data.score, None)
+        self.assertEqual(data.time_limit_msec, 2 * 1000)
+
+    # test the first format (stated in AtCoderProblemDetailedData)
+    # see https://github.com/kmyk/online-judge-tools/issues/106
+    def test_from_html_3(self):
+        url = 'https://atcoder.jp/contests/utpc2011/tasks/utpc2011_1'
+        resp = requests.get(url)
+        html = resp.content.decode(resp.apparent_encoding)
+        data = AtCoderProblemDetailedData.from_html(html, problem=AtCoderProblem.from_url(url))
+
+        self.assertEqual(data.alphabet, 'A')
+        self.assertEqual(data.available_languages, None)
+        self.assertEqual(data.html, html)
+        self.assertEqual(data.input_format, None)
+        self.assertEqual(data.memory_limit_byte, 300 * 1000 * 1000)
+        self.assertEqual(data.name, 'プログラミングコンテスト')
+        self.assertEqual(data.problem, AtCoderProblem.from_url(url))
+        self.assertEqual(data.sample_cases, [
+            TestCase(name='sample-1', input_name='入力例 1:', input_data=b'3 4\n1 0 1 0\n1 1 1 0\n0 0 0 1\n', output_name='入力例 1 に対する出力例:', output_data=b'3\n'),
+            TestCase(name='sample-2', input_name='入力例 2:', input_data=b'3 4\n1 1 1 1\n1 1 1 1\n1 1 1 1\n', output_name='入力例 2 に対する出力例:', output_data=b'4\n'),
+            TestCase(name='sample-3', input_name='入力例 3:', input_data=b'1 1\n0\n', output_name='入力例 3 に対する出力例:', output_data=b'0\n'),
+        ])
+        self.assertEqual(data.score, None)
+        self.assertEqual(data.time_limit_msec, 1 * 1000)
+
+    # deal with empty output case
+    # see https://github.com/kmyk/online-judge-tools/issues/507
+    def test_from_html_4(self):
```

Comment:
How about `test_from_html_with_empty_output`?
If this test will be failed, test function name displayed in log.
So developer can understand what happened earlier.

The case when we can't explain test contents only with function name, we should add comment.
But if not, naming test function explicitly is good practice.